### DISCUSSION TITLE
Is it even possible to target elements based on text in closed shadow root?

### DISCUSSION THREAD

**Discussion by u-RraaLL** [uBlockOrigin/uBlock-issues#2280] [Q&A]
Now that facebook started hiding their "Sponsored" text as well as all timestamps inside closed shadow root, will it be entirely impossible to target these posts?

I'm referring to [this comment](https://www.reddit.com/r/uBlockOrigin/comments/t2ymvc/comment/hysg7lg/) by @gorhill on reddit:

> Should shadow DOM be more and more used to evade cosmetic filters, I will introduce a `:shadow-root(...)` operator to deal with these cases. So far there are not enough justification to introduce this.

vs

MDN's

> When the `mode` of a shadow root is "`closed`", the shadow root's implementation internals are inaccessible and unchangeable from JavaScript—in the same way the implementation internals of, for example, the [`<video>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/video) element are inaccessible and unchangeable from JavaScript.

---

Related (although rather invalid) issue #2279

**Comment by peace2000** [comment]
FB is quite a popular page, hopefully it's a good enough justification (if it's possible to handle).

**Comment by Howking007** [comment]
Temporary solution:
`facebook.com##div.b6ax4al1:nth-of-type(19n+2)`
The sponsored posts seem to keep a fixed order (starting second and then every 19th).
This is now working for me. (but it's just a temporary solution)

**Reply by peace2000** [reply]
That matchess legit posts as well and sometimes misses sponsored, at least in the Finnish FB.

**Reply by u-RraaLL** [reply]
Somebody on reddit earlier posted a `5n+2` version. Neither one matches in my side 🤷‍♂️

**Reply by Howking007** [reply]
> Somebody on reddit earlier posted a `5n+2` version. Neither one matches in my side 🤷‍♂️

You need to count the order of the first ad (that's the 2) and then the order of the next ad (that's the 5 or 19). Find your own values.
It helps to add `:style(border: 1px dotted yellow !important)` to the end of the line. Then the given order is framed in yellow.

**Reply by u-RraaLL** [reply]
I don't need to count anything. The ads aren't injected in a uniform manner for me at all.

Disregarding the first ad (which is the 2nd post), there are 7, 7, 5, 5, 7, 3, 2, 9, 14, ..., spaces between the ads on my side now.
And if you scroll far down enough, fb will start deleting the first posts it loaded to the feed. Which means even if the numeric filter worked, it would start catching legit posts after a while.

Yes, I'm well aware how to preview cosmetic filters. I've been teaching that method to the community for years 😉

Anyway, this is not what I made this discussion for.
If you want to continue talking filter issues/ideas I invite you to do so on reddit or here https://github.com/uBlockOrigin/uAssets/issues/3367. Thanks.
